contracts/btc-light-client/src/bitcoin.rs

```diff
@@ -1,54 +1,277 @@
-                i,
-                cum_work_old + header_work,
-                cum_work,
-            ));
+        let cum_work_new = cum_work_old + block_header.work();
+        if cum_work_new != cum_work {
+            return Err(HeaderError::WrongCumulativeWork(i, cum_work_new, cum_work));
         }
-        cum_work_old = cum_work;
+
         // Validate height
         if new_header.height != last_header.height + 1 {
-            return Err(ContractError::BTCWrongHeight(
+            return Err(HeaderError::WrongHeight(
                 i,
                 last_header.height + 1,
                 new_header.height,
             ));
         }
 
         // this header is good, verify the next one
+        cum_work_old = cum_work;
         last_header = new_header.clone();
     }
     Ok(())
 }
 
+/// This functions mirrors the `checkHeader` in babylon golang implmentation.
+///
+/// https://github.com/babylonlabs-io/babylon/blob/48617fb852e9cae4ea7ea38c80793cdcb6f2668c/x/btclightclient/types/btc_light_client.go#L416
+fn check_header(
+    storage: &dyn Storage,
+    chain_params: &Params,
+    prev_block_height: u32,
+    prev_block_header: &BlockHeader,
+    header: &BlockHeader,
+) -> Result<(), HeaderError> {
+    check_block_header_context(
+        storage,
+        chain_params,
+        prev_block_height,
+        prev_block_header,
+        header,
+    )?;
+
+    // Perform proof-of-work check as specified in:
+    // https://pkg.go.dev/github.com/btcsuite/btcd@v0.24.2/blockchain#CheckBlockHeaderSanity
+    check_proof_of_work(chain_params, header)?;
+
+    Ok(())
+}
+
+/// Ensures the header's hash <= the header's target <= pow limit.
+pub(crate) fn check_proof_of_work(
+    chain_params: &bitcoin::consensus::Params,
+    header: &BlockHeader,
+) -> Result<(), HeaderError> {
+    let target = header.target();
+
+    // ensure the target <= pow_limit
+    if target > chain_params.max_attainable_target {
+        return Err(HeaderError::TargetTooLarge);
+    }
+
+    // ensure the header's hash <= target
+    // NOTE: validate_pow ensures two things
+    // - the given required_target is same
+    // - the header hash is smaller than required_target
+    // The former must be true since we give this header's target
+    // Here we are interested in the latter check, in which the code is private
+    header
+        .validate_pow(target)
+        .map_err(HeaderError::InvalidProofOfWork)?;
+
+    Ok(())
+}
+
+/// > Ensure the difficulty specified in the block header matches
+/// > the calculated difficulty based on the previous block and
+/// > difficulty retarget rules.
+///
+/// Note: While the naming mirrors the btcd implementation, this function only performs
+/// the difficulty adjustment check (not the full header context validation in btcd).
+///
+/// https://pkg.go.dev/github.com/btcsuite/btcd@v0.24.2/blockchain#CheckBlockHeaderContext
+fn check_block_header_context(
```

Comment:
the context check has also this median time check:
```go
		// Ensure the timestamp for the block header is after the
		// median time of the last several blocks (medianTimeBlocks).
		medianTime := CalcPastMedianTime(prevNode)
		if !header.Timestamp.After(medianTime) {
			str := "block timestamp of %v is not after expected %v"
			str = fmt.Sprintf(str, header.Timestamp, medianTime)
			return ruleError(ErrTimeTooOld, str)
		}
	}
```
I wonder whether missing it may lead to some discrepancies or accepting headers that should not be accepted 🤔

Reply:
Good catch! This slipped through because I got a bit mixed up between the future timestamp above and the median time check. We should definitely add this check.

Reply:
Synced with @gitferry offline, let's merge this PR on his approval for the upcoming devnet. I'll add the median time check in a follow-up PR shortly.